deepchem/models/tensorgraph/layers.py

```diff
@@ -77,30 +97,64 @@ def _create_tensor(self):
 
 class Dense(Layer):
 
-  def __init__(self, out_channels, activation_fn=None, **kwargs):
+  def __init__(
+      self,
+      out_channels,
+      activation_fn=None,
+      biases_initializer=tf.zeros_initializer,
+      weights_initializer=tf.contrib.layers.variance_scaling_initializer,
+      time_series=False,
+      scope_name=None,
+      reuse=False,
+      **kwargs):
+    super(Dense, self).__init__(**kwargs)
     self.out_channels = out_channels
     self.out_tensor = None
     self.activation_fn = activation_fn
-    super(Dense, self).__init__(**kwargs)
+    self.biases_initializer = biases_initializer
+    self.weights_initializer = weights_initializer
+    self.time_series = time_series
+    self.reuse = reuse
+    if scope_name is None:
+      scope_name = self.name
+    self.scope_name = scope_name
 
   def _create_tensor(self):
     if len(self.in_layers) != 1:
       raise ValueError("Only One Parent to Dense over %s" % self.in_layers)
     parent = self.in_layers[0]
-    if len(parent.out_tensor.get_shape()) != 2:
-      raise ValueError("Parent tensor must be (batch, width)")
-    in_channels = parent.out_tensor.get_shape()[-1].value
-    # w = initializations.glorot_uniform([in_channels, self.out_channels])
-    # w = model_ops.zeros(shape=[in_channels, self.out_channels])
-    # b = tf.Variable([0.0, 0.0])
-    # self.out_tensor = tf.matmul(parent.out_tensor, w) + b
-    self.out_tensor = tf.contrib.layers.fully_connected(
-        parent.out_tensor,
-        num_outputs=self.out_channels,
-        activation_fn=self.activation_fn,
-        scope=self.name,
-        trainable=True)
-    return self.out_tensor
+    if not self.time_series:
+      self.out_tensor = tf.contrib.layers.fully_connected(
+          parent.out_tensor,
+          num_outputs=self.out_channels,
+          activation_fn=self.activation_fn,
+          biases_initializer=self.biases_initializer(),
+          weights_initializer=self.weights_initializer(),
+          scope=self.scope_name,
+          reuse=self.reuse,
+          trainable=True)
+      return self.out_tensor
+    dense_fn = lambda x: tf.contrib.layers.fully_connected(x,
+                                                           num_outputs=self.out_channels,
+                                                           activation_fn=self.activation_fn,
+                                                           biases_initializer=self.biases_initializer(),
+                                                           weights_initializer=self.weights_initializer(),
+                                                           scope=self.scope_name,
+                                                           reuse=self.reuse,
+                                                           trainable=True)
+    self.out_tensor = tf.map_fn(dense_fn, parent.out_tensor)
+
+  def shared(self, in_layers):
```

Comment:
I think I get how this works (`tf.contrib.layers.fully_connected` looks up the variables in the provided `name_scope` and reuses them). That said, spooky action at a distance... I've never been a fan of TF's reliance on `namescopes`. Seems like a bad replacement for standard lexical scoping.

This isn't really a review comment. More me griping about Tensorflow :-)